File: src/lib/api/transactions.ts
import { apiClient } from './client';
import type { Transaction, TransactionStats } from './types';

export interface TransactionFilters {
  skip?: number;
  limit?: number;
  transaction_type?: string;
  category?: string;
  start_date?: string;
  end_date?: string;
  search?: string;
}

export const transactionsApi = {
  async getAll(filters: TransactionFilters = {}): Promise<Transaction[]> {
    const params = new URLSearchParams();
    if (filters.skip !== undefined) params.append('skip', filters.skip.toString());
    if (filters.limit !== undefined) params.append('limit', filters.limit.toString());
    if (filters.transaction_type) params.append('transaction_type', filters.transaction_type);
    if (filters.category) params.append('category', filters.category);
    if (filters.start_date) params.append('start_date', filters.start_date);
    if (filters.end_date) params.append('end_date', filters.end_date);
    if (filters.search) params.append('search', filters.search);

    const response = await apiClient.instance.get<Transaction[]>(
      `/api/transactions?${params.toString()}`
    );
    return response.data;
  },

  async getById(id: string): Promise<Transaction> {
    const response = await apiClient.instance.get<Transaction>(`/api/transactions/${id}`);
    return response.data;
  },

  async create(data: Partial<Transaction>): Promise<Transaction> {
    const response = await apiClient.instance.post<Transaction>('/api/transactions', data);
    return response.data;
  },

  async update(id: string, data: Partial<Transaction>): Promise<Transaction> {
    const response = await apiClient.instance.put<Transaction>(`/api/transactions/${id}`, data);
    return response.data;
  },

  async delete(id: string): Promise<void> {
    await apiClient.instance.delete(`/api/transactions/${id}`);
  },

  async getStats(startDate?: string, endDate?: string): Promise<TransactionStats> {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);

    const response = await apiClient.instance.get<TransactionStats>(
      `/api/transactions/stats/summary?${params.toString()}`
    );
    return response.data;
  },

  async scanReceipt(imageFile: File): Promise<Transaction> {
    const formData = new FormData();
    formData.append('file', imageFile);

    const response = await apiClient.instance.post<Transaction>(
      '/api/transactions/scan-receipt',
      formData,
      {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      }
    );
    return response.data;
  },
};

File: backend/app/api/transactions.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional
from datetime import date
from app.database import get_db
from app.models import User
from app.api.auth import get_current_user
from app.schemas.transaction import (
    TransactionCreate,
    TransactionUpdate,
    TransactionResponse,
    TransactionStats,
)
from app.services import transaction_service

router = APIRouter()


@router.get("", response_model=list[TransactionResponse])
async def get_transactions(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
    transaction_type: Optional[str] = Query(None),
    category: Optional[str] = Query(None),
    start_date: Optional[date] = Query(None),
    end_date: Optional[date] = Query(None),
    search: Optional[str] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Get all transactions with filters and pagination"""
    transactions = transaction_service.get_transactions(
        db=db,
        user_id=current_user.id,
        skip=skip,
        limit=limit,
        transaction_type=transaction_type,
        category=category,
        start_date=start_date,
        end_date=end_date,
        search=search,
    )
    return transactions


@router.get("/{transaction_id}", response_model=TransactionResponse)
async def get_transaction(
    transaction_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Get a single transaction by ID"""
    transaction = transaction_service.get_transaction(db, transaction_id, current_user.id)
    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Transaction not found"
        )
    return transaction


@router.post("", response_model=TransactionResponse, status_code=status.HTTP_201_CREATED)
async def create_transaction(
    transaction: TransactionCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Create a new transaction"""
    return transaction_service.create_transaction(db, transaction, current_user.id)


@router.put("/{transaction_id}", response_model=TransactionResponse)
async def update_transaction(
    transaction_id: str,
    transaction_update: TransactionUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Update a transaction"""
    transaction = transaction_service.update_transaction(
        db, transaction_id, current_user.id, transaction_update
    )
    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Transaction not found"
        )
    return transaction


@router.delete("/{transaction_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_transaction(
    transaction_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Delete a transaction"""
    success = transaction_service.delete_transaction(db, transaction_id, current_user.id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Transaction not found"
        )


@router.get("/stats/summary", response_model=TransactionStats)
async def get_transaction_stats(
    start_date: Optional[date] = Query(None),
    end_date: Optional[date] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Get transaction statistics"""
    return transaction_service.get_transaction_stats(
        db, current_user.id, start_date, end_date
    )

